# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df = pd.read_csv('/Users/Yasin/Desktop/Data Science Track/Guided Capstone/Step3/step3_output.csv')

In [3]:
df.head()

Unnamed: 0                           Name       state  summit_elev  \
0           2               Hilltop Ski Area      Alaska         2090   
1           4            Sunrise Park Resort     Arizona        11100   
2           5  Yosemite Ski & Snowboard Area  California         7800   
3          10               Donner Ski Ranch  California         8012   
4          12                  June Mountain  California        10090   

   vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0            294      0        0.0          0          0     0  ...   
1           1800      0        0.0          0          1     2  ...   
2            600      0        0.0          0          0     0  ...   
3            750      0        0.0          0          0     0  ...   
4           2590      0        0.0          0          2     0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              505.0       60.000000        163.000000       82.0   
4             1500.0      174.873239        115.103943       58.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0     30.000000      34.00000              152.0   
1            250.0     74.000000      78.00000              104.0   
2            300.0     47.000000      47.00000              107.0   
3            400.0     75.000000      75.00000              170.0   
4            250.0     57.916957      64.16681              128.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3             0.0         1  
4             0.0         1  

[5 rows x 27 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166 entries, 0 to 165
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         166 non-null    int64  
 1   Name               166 non-null    object 
 2   state              166 non-null    object 
 3   summit_elev        166 non-null    int64  
 4   vertical_drop      166 non-null    int64  
 5   trams              166 non-null    int64  
 6   fastEight          166 non-null    float64
 7   fastSixes          166 non-null    int64  
 8   fastQuads          166 non-null    int64  
 9   quad               166 non-null    int64  
 10  triple             166 non-null    int64  
 11  double             166 non-null    int64  
 12  surface            166 non-null    int64  
 13  total_chairs       166 non-null    int64  
 14  Runs               166 non-null    float64
 15  TerrainParks       166 non-null    float64
 16  LongestRun_mi      166 non

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [5]:
#dfo=df.select_dtypes(include=['object']) # select object type columns
dfo= pd.get_dummies(df['state'])
df_no_state = pd.concat([df, dfo], axis=1)
df_no_state = df_no_state.drop('state', axis=1)
df_no_state

Unnamed: 0                               Name  summit_elev  \
0             2                   Hilltop Ski Area         2090   
1             4                Sunrise Park Resort        11100   
2             5      Yosemite Ski & Snowboard Area         7800   
3            10                   Donner Ski Ranch         8012   
4            12                      June Mountain        10090   
..          ...                                ...          ...   
161         323                      Hogadon Basin         8000   
162         326          Sleeping Giant Ski Resort         7428   
163         327                   Snow King Resort         7808   
164         328  Snowy Range Ski & Recreation Area         9663   
165         329                White Pine Ski Area         9500   

     vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0              294      0        0.0          0          0     0       1  ...   
1             1800      0        0.0          0          1     2       3  ...   
2              600      0        0.0          0          0     0       1  ...   
3              750      0        0.0          0          0     0       1  ...   
4             2590      0        0.0          0          2     0       0  ...   
..             ...    ...        ...        ...        ...   ...     ...  ...   
161            640      0        0.0          0          0     0       0  ...   
162            810      0        0.0          0          0     0       1  ...   
163           1571      0        0.0          0          0     1       1  ...   
164            990      0        0.0          0          0     0       1  ...   
165           1100      0        0.0          0          0     0       2  ...   

     Rhode Island  South Dakota  Tennessee  Utah  Vermont  Virginia  \
0               0             0          0     0        0         0   
1               0             0          0     0        0         0   
2               0             0          0     0        0         0   
3               0             0          0     0        0         0   
4               0             0          0     0        0         0   
..            ...           ...        ...   ...      ...       ...   
161             0             0          0     0        0         0   
162             0             0          0     0        0         0   
163             0             0          0     0        0         0   
164             0             0          0     0        0         0   
165             0             0          0     0        0         0   

     Washington  West Virginia  Wisconsin  Wyoming  
0             0              0          0        0  
1             0              0          0        0  
2             0              0          0        0  
3             0              0          0        0  
4             0              0          0        0  
..          ...            ...        ...      ...  
161           0              0          0        1  
162           0              0          0        1  
163           0              0          0        1  
164           0              0          0        1  
165           0              0          0        1  

[166 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [6]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df_no_state.drop(['Name', 'AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df_no_state['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [7]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [8]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score, mean_absolute_error
lm = linear_model.LinearRegression()
lm.fit(X_train,y_train)


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [9]:
lm.score(X_train,y_train)

0.8912046133387441

In [10]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = lm.predict(X_test)
print(lm.score(X_test, y_test))

-1.4538284124034147e+22


## Review Model Outcomes — Iterate over additional models as needed

In [11]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
#from sklearn.metrics import explained_variance_score, mean_absolute_error
print(explained_variance_score(y_test, y_pred))

-1.419213450203144e+22


In [12]:
print(mean_absolute_error(y_test, y_pred))

276268327411.36346


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [13]:
lm.intercept_

69899215424.45668

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
Unnamed: 0         1.615676e+01
summit_elev        1.572105e+00
vertical_drop      1.132705e+00
trams              6.270245e+11
fastEight          3.112477e+13
fastSixes          1.195909e+13
fastQuads          1.677182e+12
quad               2.106351e+12
triple             3.483641e+12
double             4.070812e+12
surface            4.224922e+12
total_chairs       7.384853e+12
Runs               6.826172e-01
TerrainParks       1.051758e+00
LongestRun_mi      7.109375e-01
SkiableTerrain_ac  8.208008e-01
Snow Making_ac     9.111328e-01
daysOpenLastYear   1.545166e+00
yearsOpen          6.791992e-01
averageSnowfall    4.853516e-01
AdultWeekday       8.944336e+00
projectedDaysOpen  2.488770e+00
NightSkiing_ac     1.291016e+00
clusters           3.916016e-01
Alaska             8.978717e+11
Arizona            8.978717e+11
California         2.485111e+12
Colorado           1.983217e+12
Connecticut        1.983217e+12
Idaho              2.165748e+12
Illinois           1.545705e+12
Indiana            1.265929e+12
Iowa               1.265929e+12
Maine              1.983217e+12
Maryland           8.978717e+11
Massachusetts      1.779344e+12
Michigan           3.326645e+12
Minnesota          2.485111e+12
Missouri           1.265929e+12
Montana            1.545705e+12
Nevada             8.978717e+11
New Hampshire      2.886254e+12
New Jersey         8.978717e+11
New Mexico         2.165748e+12
New York           3.694093e+12
North Carolina     1.545705e+12
Ohio               1.779344e+12
Oregon             1.545705e+12
Pennsylvania       2.760798e+12
Rhode Island       8.978717e+11
South Dakota       2.112296e+03
Tennessee          8.978717e+11
Utah               1.545705e+12
Vermont            2.331953e+12
Virginia           1.265929e+12
Washington         1.779344e+12
West Virginia      1.545705e+12
Wisconsin          2.760798e+12
Wyoming            1.983217e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [15]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend', and 'state' from the df
X = df.drop(['Name', 'AdultWeekend', 'state'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

In [16]:
y = y.ravel()

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [17]:
lm = linear_model.LinearRegression()
lm.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [18]:
print(lm.score(X_train,y_train))

0.7937629274542103


In [19]:
y_pred = lm.predict(X_test)

In [20]:
print(explained_variance_score(y_test, y_pred))

0.7838124661158787


In [21]:
print(mean_absolute_error(y_test, y_pred))

5.619585726667313


In [22]:
lm.intercept_

56.65098787380902

In [23]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0         5.199001e-01
summit_elev        3.259379e+00
vertical_drop      1.981040e+00
trams              2.664535e-15
fastEight          2.220446e-16
fastSixes          4.440892e-16
fastQuads          9.845029e-02
quad               1.728484e+00
triple             8.877445e-01
double             1.203186e+00
surface            1.387656e+00
total_chairs       8.034957e-01
Runs               6.887657e-01
TerrainParks       8.305176e-01
LongestRun_mi      6.328773e-01
SkiableTerrain_ac  4.461051e-01
Snow Making_ac     9.050997e-01
daysOpenLastYear   2.162870e+00
yearsOpen          1.727835e-01
averageSnowfall    1.243931e+00
AdultWeekday       1.086174e+01
projectedDaysOpen  1.984044e+00
NightSkiing_ac     3.844018e-01
clusters           2.391744e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [24]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend', and 'state' from the df
X = df.drop(['Name', 'AdultWeekend', 'state', 'summit_elev'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

In [25]:
y = y.ravel()

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [26]:
lm = linear_model.LinearRegression()
lm.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [27]:
print(lm.score(X_train,y_train))

0.7796964779269345


In [28]:
y_pred = lm.predict(X_test)

In [29]:
print(explained_variance_score(y_test, y_pred))

0.7684059981059617


In [30]:
print(mean_absolute_error(y_test, y_pred))

5.854793366890659


In [31]:
lm.intercept_

56.700123302497964

In [32]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0         1.547955e-01
vertical_drop      1.714111e+00
trams              2.664535e-15
fastEight          1.332268e-15
fastSixes          1.110223e-15
fastQuads          1.669965e-01
quad               1.447971e+00
triple             7.980164e-01
double             1.374414e+00
surface            1.042259e+00
total_chairs       9.887179e-01
Runs               7.162610e-01
TerrainParks       1.104044e+00
LongestRun_mi      9.112824e-01
SkiableTerrain_ac  1.353448e+00
Snow Making_ac     1.035596e+00
daysOpenLastYear   1.570286e+00
yearsOpen          9.092574e-02
averageSnowfall    2.160135e+00
AdultWeekday       1.058029e+01
projectedDaysOpen  1.540808e+00
NightSkiing_ac     9.165926e-01
clusters           1.814971e+00

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.419213450203144e+22 | 276268327411.36346 |-|
| Model 2. | 0.7838124661158787| 5.619585726667313 |'state'|
| Model 3. | 0.7684059981059617 | 5.854793366890659 |'state','summit_elev','base_elev'|

Model Selection:

Based on these 3 models, the best one is Model 2.